core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -178,6 +179,34 @@ inline void configure_shmem_preference(KernelFuncPtr const& func,
 #endif
 }
 
+template <class Policy>
+std::enable_if_t<Policy::experimental_contains_desired_occupancy>
+modify_launch_configuration_if_desired_occupancy_is_specified(
+    Policy const& policy, cudaDeviceProp const& properties,
+    cudaFuncAttributes const& attributes, dim3 const& block, int& shmem,
+    bool& prefer_shmem) {
+  int const block_size        = block.x * block.y * block.z;
+  int const desired_occupancy = policy.impl_get_desired_occupancy().value();
+
+  size_t const shmem_per_sm_prefer_l1 = get_shmem_per_sm_prefer_l1(properties);
+  size_t const static_shmem           = attributes.sharedSizeBytes;
+  int active_blocks = properties.maxThreadsPerMultiProcessor / block_size *
+                      desired_occupancy / 100;
+  size_t const dynamic_shmem =
+      shmem_per_sm_prefer_l1 / active_blocks - static_shmem;
+
+  if ((int)dynamic_shmem > shmem) {
```

Comment:
Do we still want C-style casts?  (see P/R #3522)

Reply:
Fixed